Add Druid integration (#4070)
DataDog/integrations-core: druid/README.md

# Agent Check: druid

## Overview

> The Datadog Agent collects many metrics from _Druid_ via DogStatsD. Those Druid metrics are related to queries, ingestion, and coordination:

Reviewer: ```suggestion
The Datadog Agent collects metrics from Druid via [DogStatsD](https://docs.datadoghq.com/developers/dogstatsd/). Metrics collected are related to Druid queries, ingestion, and coordination data, see the [Druid metrics documentation][1] to learn more.
```

> * Query Metrics

Reviewer: Since we refer the reader to the metric section, this list can be removed

Author: fixed

* SQL Metrics
* Coordination
* General Health
* Sys

And many more. See details [here][1].

> I addition to metrics, the Datadog Agent also sends a service check for Druid health.

Reviewer: ```suggestion
I addition to metrics collected, the Agent also sends a Service Check related to Druid's health.
```

## Setup

### Prerequisite

Druid 0.16 or above is required for this integration to work properly.

### Installation

> The druid check is included in the [Datadog Agent][2] package.

Reviewer: ```suggestion
The Druid check is included in the [Datadog Agent][2] package.
```

Author: fixed

> No additional installation is needed on your server.

Reviewer 2: Not true, as per line 38

Reviewer: agreed, this sentence should be removed and instructions starting at line 38 should be inlined here since they are for installation process.

Author: I updated the doc for `Installation`. Could you have a look again. Thx

1) Install Druid extension [`statsd-emitter`][5] 

```
$ java \
  -cp "lib/*" \
  -Ddruid.extensions.directory="./extensions" \
  -Ddruid.extensions.hadoopDependenciesDir="hadoop-dependencies" \
  org.apache.druid.cli.Main tools pull-deps \
  --no-default-hadoop \
  -c "org.apache.druid.extensions.contrib:statsd-emitter:0.15.0-incubating"
```

More info about this step can be found on the [official guide for loading Druid extensions][6]

2) Update Druid java properties

Update/Add following configs to your druid properties.
```
# Add `statsd-emitter` to the extensions list to be loaded 
druid.extensions.loadList=[..., "statsd-emitter"]

# By default druid emission period is 1 minute (PT1M).
# We recommmend using 15 seconds instead as follow:
druid.monitoring.emissionPeriod=PT15S

# Use `statsd-emitter` extension as metric emitter
druid.emitter=statsd

# Configure `statsd-emitter` endpoint
druid.emitter.statsd.hostname=127.0.0.1
druid.emitter.statsd.port:8125

# Configure `statsd-emitter` to use dogstatsd format 
druid.emitter.statsd.dogstatsd=true
druid.emitter.statsd.dogstatsdServiceAsTag=true
```

> Important: `druid.emitter.statsd.dogstatsd` and `druid.emitter.statsd.dogstatsdServiceAsTag` must be set to `true`, otherwise tags might not be reported correctly to Datadog.

Reviewer: ```suggestion
**Note**: `druid.emitter.statsd.dogstatsd` and `druid.emitter.statsd.dogstatsdServiceAsTag` must be set to `true`, otherwise tags are not reported correctly to Datadog.
```

> Restart Druid to start sending more Druid metrics to DogStatsD (included in the Datadog Agent).

Reviewer: ```suggestion
Restart Druid to start sending your Druid metrics to the Agent through DogStatsD.
```

#### Integration Service Checks

Use the default configuration of your `druid.d/conf.yaml` file to activate the collection of your Druid service checks. See the sample [druid.d/conf.yaml][3] for all available configuration options.

### Validation

[Run the Agent's status subcommand][7] and look for `druid` under the Checks section.

## Data Collected

### Metrics

See [metadata.csv][8] for a list of metrics provided by this check.

### Service Checks

**druid.process.can_connect**:

Returns `CRITICAL` if the check cannot connect to Druid process. Returns `OK` otherwise.

**druid.process.health**:

Returns `CRITICAL` if Druid process is not healthy. Returns `OK` otherwise.

> 

Reviewer: extra line break not necessary

Author: fixed

# Agent Check: druid

(TBD Add Druid Dashboard Image here)

## Overview

The Datadog Agent collects many metrics from _Druid_ via DogStatsD. Those Druid metrics are related to queries, ingestion, and coordination:

* Query Metrics
* SQL Metrics
* Coordination
* General Health
* Sys

And many more. See details [here][1].

I addition to metrics, the Datadog Agent also sends a service check for Druid health.

## Setup

### Prerequisite

> Druid 0.16 or above is required for this integration to work properly.

Author: Druid 0.16 is not release yet, but should be soon (maybe sept 2019). Druid Integration should not be public until it's released, it contains upstream improvements for Druid `statsd-emitter` extension.
